### PR TITLE
Fix Unicode

### LINKED ISSUE #123: thread 'main' panicked at 'internal error: entered unreachable code', helix-core/src/movement.rs:203:9
I'm running on commit 407b37c3279bfd0ae2bf756bc022d47d5db446d9

I got this error when writing some Japanese text in my buffer. Issue is likely with the character `。`. I changed code to display the character, here is the full trace:
```
thread 'main' panicked at 'internal error: entered unreachable code: '。'', helix-core/src/movement.rs:203:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
   2: helix_core::movement::categorize
             at ./helix-core/src/movement.rs:203:9
   3: helix_core::movement::move_prev_word_start
             at ./helix-core/src/movement.rs:113:30
   4: hx::commands::move_prev_word_start::{{closure}}
             at ./helix-term/src/commands.rs:241:9
   5: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:280:13
   6: core::option::Option<T>::map
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:487:29
   7: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:101:9
   8: <smallvec::SmallVec<A> as core::iter::traits::collect::Extend<<A as smallvec::Array>::Item>>::extend
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-1.6.1/src/lib.rs:1663:36
   9: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-1.6.1/src/lib.rs:1648:9
  10: core::iter::traits::iterator::Iterator::collect
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1764:9
  11: helix_core::selection::Selection::transform
             at ./helix-core/src/selection.rs:281:13
  12: hx::commands::move_prev_word_start
             at ./helix-term/src/commands.rs:240:21
  13: hx::ui::editor::EditorView::command_mode
             at ./helix-term/src/ui/editor.rs:530:21
  14: <hx::ui::editor::EditorView as hx::compositor::Component>::handle_event
             at ./helix-term/src/ui/editor.rs:618:33
  15: hx::compositor::Compositor::handle_event
             at ./helix-term/src/compositor.rs:112:19
  16: hx::application::Application::handle_terminal_events
             at ./helix-term/src/application.rs:144:32
  17: hx::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:108:21
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  19: hx::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:266:9
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  21: hx::main::{{closure}}
             at ./helix-term/src/main.rs:157:5
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  23: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:54
  24: tokio::coop::with_budget::{{closure}}
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:106:9
  25: std::thread::local::LocalKey<T>::try_with
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  26: std::thread::local::LocalKey<T>::with
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  27: tokio::coop::with_budget
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:99:5
  28: tokio::coop::budget
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:76:5
  29: tokio::park::thread::CachedParkThread::block_on
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:31
  30: tokio::runtime::enter::Enter::block_on
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/enter.rs:151:13
  31: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/thread_pool/mod.rs:71:9
  32: tokio::runtime::Runtime::block_on
             at /home/auroden/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/mod.rs:452:43
  33: hx::main
             at ./helix-term/src/main.rs:159:5
  34: core::ops::function::FnOnce::call_once
             at /home/auroden/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
```

UPDATE: it happens when moving using commands like `b`, `e` or `w` with various unicode characters (another one is `…`).

### PR DESCRIPTION
**Pull request by kirawi**
Fix #123, should probably still include #129 as a fallback. Remember how I said there shouldn't be any regressions in #121? Well, I figured out my "fix" for #88 broke `b` on unicode especially, but now I guess that's fixed here since it properly detects unicode?

### CODE REVIEW

**Comment by kirawi** [comment]
Oops, accidentally included old commits.

File: helix-core/src/transaction.rs [hunk 1]
```diff
@@ -754,4 +755,21 @@ mod test {
         use Operation::*;
         assert_eq!(changes.changes, &[Insert("a".into())]);
     }
+
+    #[test]
+    fn combine_with_utf8() {
+        const TEST_CASE: &'static str = "Hello, これはヒレクスエディターです！";
```
Comment (CBenoit) [inline helix-core/src/transaction.rs:761]: Unrelated to code, but I expected ヘリックス here :laughing: 
```suggestion
        const TEST_CASE: &'static str = "Hello, これはヘリックスエディターです！";
```

Comment (kirawi) [reply helix-core/src/transaction.rs:761]: My 日本語 is not 上手 ;-;

File: helix-core/src/movement.rs [hunk 2]
```diff
@@ -192,11 +204,11 @@ enum Category {
 fn categorize(ch: char) -> Category {
     if ch == '\n' {
         Category::Eol
-    } else if ch.is_ascii_whitespace() {
```
Comment (archseer) [inline helix-core/src/movement.rs:195]: I think this could simply use char.is_whitespace, which is included in stdlib

Comment (archseer) [reply helix-core/src/movement.rs:195]: Alternatively we can call get_general_category once then switch based on the result

Comment (kirawi) [reply helix-core/src/movement.rs:195]: It breaks with unicode space, i.e. '　' in Japanese.

Comment (archseer) [reply helix-core/src/movement.rs:195]: is_ascii_whitespace does, but is_whitespace works with unicode: https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace

Comment (kirawi) [reply helix-core/src/movement.rs:195]: Gotcha

Comment (kirawi) [reply helix-core/src/movement.rs:195]: Would it be worth using it over the one in the match?

**Comment by kirawi** [comment]
Updated

**Comment by kirawi** [comment]
Also, I just found out that it doesn't fix the unicode problem regex problem exactly, I'll look into it tomorrow.

**Comment by archseer** [comment]
I think it looks good, but what is '_' categorized as now?

**Comment by kirawi** [comment]
Added a test, and it's still recognized as a word. The categories can be seen here, https://www.compart.com/en/unicode/category. I'll look into the aforementioned regression tomorrow, or possibly later depending on my schedule.

File: helix-core/src/transaction.rs [hunk 3]
```diff
@@ -758,7 +758,7 @@ mod test {
 
     #[test]
     fn combine_with_utf8() {
-        const TEST_CASE: &'static str = "Hello, これはヒレクスエディターです！";
+        const TEST_CASE: &'static str = "Hello, これはヘリックスエディターです！";
```
Comment (pickfire) [inline helix-core/src/transaction.rs:761]: @CBenoit Err, why isn't it `へニックス`
```suggestion
        const TEST_CASE: &'static str = "Hello, これはへニックスエディターです！";
```

File: helix-core/src/movement.rs [hunk 4]
```diff
@@ -190,16 +196,22 @@ enum Category {
...
+    }
+
+    match get_general_category(ch) {
+        GeneralCategory::SpaceSeparator => Category::Whitespace,
+        GeneralCategory::OtherPunctuation
+        | GeneralCategory::ClosePunctuation
+        | GeneralCategory::OpenPunctuation
+        | GeneralCategory::InitialPunctuation
+        | GeneralCategory::FinalPunctuation => Category::Punctuation,
+        _ => unreachable!("unknown '{}' character category", ch),
```
Comment (pickfire) [inline helix-core/src/movement.rs:214]: This will fail for other cases, like `+` it will get a panic.

File: helix-core/src/movement.rs [hunk 5]
```diff
@@ -257,4 +269,25 @@ mod test {
             (1, 2).into()
         );
     }
+
+    #[test]
+    fn test_categorize() {
+        const WORD_TEST_CASE: &'static str = "_hello_world_あいうえお";
+        const PUNCTUATION_TEST_CASE: &'static str = ".,!?;:。、！？；：{}[]｛｝「」⟪";
+        const WHITESPACE_TEST_CASE: &'static str = "  　   ";
```
Comment (pickfire) [inline helix-core/src/movement.rs:277]: Math symbols fail, should we add a test case? `+-*/`

**Comment by kirawi** [comment]
Might be worth having a separate test containing the equivalent of, "the fox jumped over the... Etc.", though I have no idea what that would be for Unicode. Maybe foreign equivalents? The goal of it would be to check for any panics.

**Comment by pickfire** [comment]
> Might be worth having a separate test containing the equivalent of, "the fox jumped over the... Etc.", though I have no idea what that would be for Unicode. Maybe foreign equivalents? The goal of it would be to check for any panics. I guess we should just get one character from each category.

If you want that, fuzzy will be easier since it spams random characters, then we run our code through the fuzzer.

**Comment by archseer** [comment]
Property testing: https://github.com/BurntSushi/quickcheck

File: helix-core/src/transaction.rs (hunk 3 above)
Comment (kirawi) [reply helix-core/src/transaction.rs:761]: ニ is ni, リ is ri 😅

**Comment by CBenoit** [comment]
I changed a bit #129 to return a `Unknown` variant instead. Your changes should be enough to handle most useful unicodes.

**Comment by kirawi** [comment]
> I changed a bit #129 to return a `Unknown` variant instead. Your changes should be enough to handle most useful unicodes.

That's perfect, I was a little worried about how I could create tests for this PR.

**Comment by pickfire** [comment]
@kirawi Can you please rebase and add it for ctrl-w in words.rs as well?

**Comment by kirawi** [comment]
#88 was not fixed, that issue should be reopened. I still have no idea what is causing the bug.

**Comment by kevinsjoberg** [comment]
@kirawi actually the test case in #83 is fixed. I guess what I'm seeing now could count as a new issue. I'll create one!

**Comment by kirawi** [comment]
> @kirawi actually the test case in #88 is fixed. I guess what I'm seeing now could count as a new issue. I'll create one!

Are you sure? I noticed that it chokes if there is more than one unicode, i.e.
```
key
ささ
key
key
```

**Comment by kevinsjoberg** [comment]
Ah, interesting. Never tried that. Perhaps that's why it chokes on the new input. Well, I'll add a comment t to the original issue tomorrow. Getting late here.

File: helix-core/src/words.rs [hunk 6]
```diff
@@ -47,11 +47,11 @@ fn different_prev_word_boundary() {
...
+    t("こんにちは、世界！", "こんにちは、世界");
     t("こんにちは、世界", "こんにちは、");
-    t("こんにちは、", "こんにちは、"); // what?
+    t("こんにちは、", "こんにちは");
     t("こんにちは", "");
-    t("この世界。", "この世界。"); // what?
+    t("この世界。", "この世界");
     t("この世界", "");
     t("お前はもう死んでいる", "");
     t("その300円です", ""); // TODO: should stop at 300
```
Comment (kirawi) [inline helix-core/src/words.rs:57]: Unrelated to the PR, but https://github.com/PSeitz/rust-tinysegmenter might be possible here.

**Review by archseer — APPROVED** [review]

File: helix-core/src/words.rs (hunk 6 above)
Comment (pickfire) [reply helix-core/src/words.rs:57]: Nah, I think this is fine already once it can handle the `。`.

File: helix-core/src/movement.rs [hunk 7]
```diff
@@ -174,12 +174,28 @@ pub fn move_next_word_end(slice: RopeSlice, mut begin: usize, count: usize) -> O
...
+            | GeneralCategory::ClosePunctuation
+            | GeneralCategory::InitialPunctuation
+            | GeneralCategory::FinalPunctuation
+            | GeneralCategory::ConnectorPunctuation
+            | GeneralCategory::DashPunctuation
+            | GeneralCategory::MathSymbol
+            | GeneralCategory::CurrencySymbol
+            | GeneralCategory::ModifierSymbol
+    )
 }
```
Comment (pickfire) [inline helix-core/src/movement.rs:199]: I think we should move these `pub(crate)` and categories stuff to `category.rs` or something. Maybe in another PR.

File: helix-core/src/movement.rs [hunk 8]
```diff
@@ -231,6 +249,7 @@ where
     // need to +1 so that prev() includes current char
     let mut chars = slice.chars_at(*pos + 1);
 
+    #[allow(clippy::while_let_on_iterator)]
```
Comment (pickfire) [inline helix-core/src/movement.rs:252]: Is this a false positive, maybe report it to clippy?

**Review by pickfire — APPROVED** [review]
Looks good to me. Merging now but we need some follow up.

File: helix-core/src/movement.rs (hunk 8 above)
Comment (kirawi) [reply helix-core/src/movement.rs:252]: Not exactly, but the next line operates on next(), and I didn't know how to get with that so I decided to surprise the warning for now.

**Comment by archseer** [comment]
Squash and merge works

**Pull request merged**